Sprint_14

MiraiChan opened the pull request:
Sprint_14 done.

MiraiChan committed:
Launch screen was set.

MiraiChan committed:
TabBarController was added.
Relevant color sets and icons were uploaded.

MiraiChan committed:
The main screen "TrackersViewController" was partially coded:
1) the title "Trackers" was added;
2) + button in the left NavBarItem to add new trackers was added;
3) a stub with question were added;
4) empty method didAddTrackerButtonTapped() was added.

MiraiChan committed:
"Plus" icon was added according to Figma prototype.

MiraiChan committed:
DateFormatter, UIDatePicker and SearchBar were added.

MiraiChan committed:
TrackersViewController was done:
method setupAddTrackerButton() was changed (constraints)
UIStackView was added
New class NewTrackerViewController() was prepared for filling with future code.

MiraiChan committed:
Models were added and configured.

MiraiChan committed:
Main storyboard was deleted.
NewTrackerViewController was created and set.

MiraiChan committed:
TrackersViewController was restructured and UICollectionView was added.
UIDatePicker() was added to the upper right corner as a NavBarItem.
TrackerCell was created and partially set.
First Tracker creation screen was implemented: NewTrackerViewController.
Second Tracker creation screen was implemented when Habit selected: NewHabitViewController.
Color extension was added to convert to rgba format.
UICollectionView was added to etrieve a reusable cell from a collection.
TrackersStorage and ServiceLocator as a singleton for accessing to the storage of trackers were added.
Method getWeekDay() to convert a numeric day of the week value into enumeration value was implemented.
TrackerCell and TrackerTitleSectionView were added to form a card of habits.

MiraiChan committed:
Tracker Cell was fully changed.

MiraiChan committed:
TrackersViewController was restructured again after fixing issues with constraints and configuring Collection View.
Some file were deleted due to excessiveness.
New color sets and images were added to assets.
TrackerTitleSectionView was restructured after setting new constrains and deleting some methods.
TrackersStorage was simplified.
NewTrackerViewController and NewHabitViewControllers were fully changed.

MiraiChan committed:
ScheduleViewController and ScheduleViewCell were done.

MiraiChan committed:
The addButton was set according to the prototype after tap action.
Chevron icon was added to assets.
dateFormat = "dd.MM.yy" was added for making changes later.

Helvetii951 commented on Tracker/Application/AppDelegate.swift:
```diff
@@ -0,0 +1,59 @@
+    
+    // MARK: - Core Data stack
+    
+    lazy var persistentContainer: NSPersistentContainer = {
+        let container = NSPersistentContainer(name: "Tracker")
+        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
+            if let error = error as NSError? {
+                fatalError("Unresolved error \(error), \(error.userInfo)")
+            }
+        })
+        return container
+    }()
+    
+    // MARK: - Core Data Saving support
+    
+    func saveContext () {
+        let context = persistentContainer.viewContext
+        if context.hasChanges {
+            do {
+                try context.save()
+            } catch {
+                let nserror = error as NSError
+                fatalError("Unresolved error \(nserror), \(nserror.userInfo)")
+            }
+        }
+    }
```
Не совсем понятно зачем здесь используется CoreData. Она будет только в следующих принтах.

MiraiChan replied:
Убрала.

Helvetii951 commented on Tracker/Application/AppDelegate.swift:
```diff
@@ -0,0 +1,59 @@
+        if context.hasChanges {
+            do {
+                try context.save()
+            } catch {
+                let nserror = error as NSError
+                fatalError("Unresolved error \(nserror), \(nserror.userInfo)")
+            }
```
Рекомендую избегать `fatalError`. Тоже может привести к неприятным падениям в продакшене. Лучше использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfailure(_:file:line:)?changes=latest__7).

MiraiChan replied:
Ок.

Helvetii951 commented on Tracker/Application/ServiceLocator.swift:
```diff
@@ -0,0 +1,21 @@
+    
+    private init() {
+        trackerStorage = TrackersStorage.shared
+    }
```
Зависимость `trackerStorage: TrackersStorageProtocol` по протоколу. А создается в ините конкретная реализация. 

[Статья по внедрению завимостей](https://habr.com/ru/companies/tinkoff/articles/546360/)

MiraiChan replied:
При коммите файлы, которые я давно удалила/убрала референс все равно попали на гит, сейчас их уже не существует, но при пул реквесте я этого не заметила. У меня давно этого нет и проект билдится и все показывает. То, что тут вы закомментировали этого уже нет, это были мои эксперименты. Я их удалила давно.

Helvetii951 commented on Tracker/Cells/NewHabitTrackerViewCell.swift:
```diff
@@ -0,0 +1,54 @@
+
+final class NewHabitTrackerViewCell: UITableViewCell {
+    
+    private let titleLabel: UILabel = {
+        let label = UILabel()
+        label.font = UIFont.systemFont(ofSize: 16, weight: .regular)
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
+    
+    private let chevronImage: UIImageView = {
+        let chevronImage = UIImageView()
+        chevronImage.image = UIImage(named: "Chevron")
+        chevronImage.tintColor = .ypGray
+        chevronImage.translatesAutoresizingMaskIntoConstraints = false
+        return chevronImage
+    }()
```
Рекомендую использовать [ленивые свойства ](https://abhimuralidharan.medium.com/lazy-var-in-ios-swift-96c75cb8a13a)для объявления подобных свойств.  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память.

MiraiChan replied:
Пофиксила.

MiraiChan committed:
Deleted unnecessary files.

MiraiChan commented:
У меня столько замечаний по работе: того нет , этого нет. Я не знаю почему, но у меня все присутствует. Ваши скрины не соответствуют с моей работой. Я сделала пуш, и почистила ненужные файлы , которые все-таки как-то попали на гит. Надеюсь, во второй раз откроется нормально проект. 
<img width="381" alt="Screenshot 2023-11-20 at 07 44 55" src="https://github.com/MiraiChan/Tracker/assets/39549759/3a5299fb-0f39-4c7f-a80d-74eb7462255f">
<img width="381" alt="Screenshot 2023-11-20 at 07 45 03" src="https://github.com/MiraiChan/Tracker/assets/39549759/7904618d-0580-41a9-835a-3d04b0ce6c56">

Helvetii951 commented:
> У меня столько замечаний по работе: того нет , этого нет. Я не знаю почему, но у меня все присутствует. Ваши скрины не соответствуют с моей работой. Я сделала пуш, и почистила ненужные файлы , которые все-таки как-то попали на гит. Надеюсь, во второй раз откроется нормально проект. <img alt="Screenshot 2023-11-20 at 07 44 55" width="381" src="https://user-images.githubusercontent.com/39549759/284155197-3a5299fb-0f39-4c7f-a80d-74eb7462255f.png"> <img alt="Screenshot 2023-11-20 at 07 45 03" width="381" src="https://user-images.githubusercontent.com/39549759/284155211-7904618d-0580-41a9-835a-3d04b0ce6c56.png">

Сейчас скачал твой код, еще раз проверил. Залил видео на ютуб. [Проблемы остались](https://youtube.com/shorts/SB_cehCUMjA).

MiraiChan committed:
Added "lazy", changed fatalErrors to assertionFailure.
Deleted unrelated code from AppDelegate and SceneDelegate files.
Added two new files: IrregularViewController and IrregularEventCell.

MiraiChan commented:
А, поняла, спасибо. Подебажу.

MiraiChan committed:
Fixed duplicated trackers. Trackers are send to temporary mock category.
Added separator to TabBar.
Fixed search.

MiraiChan commented:
Пока файл с категориями не создан, временно добавлю трекер в конкретную категорию, чтобы не было дублирования.
